server/jobs/legal_hold_job.go

```diff
@@ -186,49 +215,64 @@ func (j *LegalHoldJob) run() {
 	settings = j.settings.Clone()
 	j.mux.Unlock()
 
-	// Retrieve the legal holds from the store.
-	legalHolds, err := j.kvstore.GetAllLegalHolds()
-	if err != nil {
-		j.client.Log.Error("Failed to fetch legal holds from store", err)
-	}
+	for _, lh := range legalHolds {
+		now := mattermostModel.GetMillis()
 
-	j.client.Log.Info("Processing all Legal Holds", "count", len(legalHolds))
+		legalHold := lh.DeepCopy()
 
-	for _, lh := range legalHolds {
 		for {
-			if lh.IsFinished() {
-				j.client.Log.Debug(fmt.Sprintf("Legal Hold %s has ended and therefore will not execute.", lh.ID))
+			if legalHold.IsFinished() {
+				j.client.Log.Debug(fmt.Sprintf("Legal Hold %s has ended and therefore does not executing.", legalHold.ID))
 				break
 			}
 
-			if !lh.NeedsExecuting(mattermostModel.GetMillis()) {
-				j.client.Log.Debug(fmt.Sprintf("Legal Hold %s is not yet ready to be executed again.", lh.ID))
+			if !forceRun && !legalHold.NeedsExecuting(now) {
+				j.client.Log.Debug(fmt.Sprintf("Legal Hold %s is not yet ready to be executed again.", legalHold.ID))
+				break
+			}
+			if legalHold.LastExecutionEndedAt >= now {
+				j.client.Log.Debug(fmt.Sprintf("Legal Hold %s was already executed after the current time.", legalHold.ID))
 				break
 			}
 
-			j.client.Log.Debug(fmt.Sprintf("Creating Legal Hold Execution for legal hold: %s", lh.ID))
-			lhe := legalhold.NewExecution(lh, j.papi, j.sqlstore, j.filebackend)
+			j.client.Log.Debug(fmt.Sprintf("Creating Legal Hold Execution for legal hold: %s", legalHold.ID))
+			lhe := legalhold.NewExecution(legalHold, j.papi, j.sqlstore, j.kvstore, j.filebackend)
 
-			if end, err := lhe.Execute(); err != nil {
+			if updatedLH, err := lhe.Execute(now); err != nil {
+				if strings.Contains(err.Error(), "another execution is already running") {
```

Comment:
Is `another execution is already running` still the error string returned when the mutex cannot be locked?

Reply:
I don't know where this comes from. At first I though this was added by aider somehow but I added this myself and it just couldn't work because the cluster mutex does not return an error if the mutex is already locked, it waits for it to be unlocked.